src/System/refine.rs

```diff
@@ -4,9 +4,79 @@ use log::{debug, info, log_enabled, trace, warn, Level};
 use crate::DataTypes::{PassedStateList, PassedStateListExt, WaitingStateList};
 use crate::ModelObjects::component::Transition;
 
+use crate::extract_system_rep::SystemRecipeFailure;
 use crate::ModelObjects::statepair::StatePair;
-use crate::TransitionSystems::{LocationTuple, TransitionSystemPtr};
+use crate::System::local_consistency::ConsistencyFailure;
+use crate::TransitionSystems::common::CollectionOperation;
+use crate::TransitionSystems::transition_system::PrecheckResult;
+use crate::TransitionSystems::{LocationID, LocationTuple, TransitionSystemPtr};
 use std::collections::HashSet;
+use std::fmt;
+
+/// The result of a refinement check. [RefinementFailure] specifies the failure.
+#[allow(clippy::large_enum_variant)] //TODO: consider boxing the large fields to reduce the total size of the enum
+pub enum RefinementResult {
+    Success,
+    Failure(RefinementFailure),
+}
+
+/// The failure of a refinement check. Variants with [StatePair] include the
+/// state that caused the failure. Variants with [LocationID] include the
+/// locations that caused failure.
+#[derive(Debug)]
+pub enum RefinementFailure {
+    NotDisjointAndNotSubset(SystemRecipeFailure),
+    NotDisjoint(SystemRecipeFailure),
+    NotSubset,
+    CutsDelaySolutions(StatePair),
+    InitialState(StatePair),
+    EmptySpecification,
+    EmptyImplementation,
+    EmptyTransition2s(StatePair),
+    NotEmptyResult(StatePair),
+    ConsistencyFailure(Option<LocationID>, Option<String>),
+    DeterminismFailure(Option<LocationID>, Option<String>),
+}
```

Comment:
Should probably be split into sub-enums (failing on preconditions, failing on initial state, failing during check), given more descriptive names and error messages, `CutsDelaySolutions`, `EmptyTransition2s` and `NotEmptyResult` can and should also include the action that caused the error.